File: android-app/app/src/main/java/com/ufscar/dc/movel/walletapp/ui/compose/MainScreen.kt
package com.ufscar.dc.movel.walletapp.ui.compose

import androidx.compose.foundation.layout.*
import androidx.compose.foundation.lazy.LazyColumn
import androidx.compose.foundation.lazy.items
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.ShoppingCart
import androidx.compose.material3.*
import androidx.compose.runtime.*
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.Color
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.ui.res.stringResource
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.style.TextAlign
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import androidx.lifecycle.viewmodel.compose.viewModel
import com.ufscar.dc.movel.walletapp.R
import com.ufscar.dc.movel.walletapp.ui.theme.WalletAppTheme
import androidx.compose.ui.tooling.preview.Preview
import java.text.NumberFormat
import java.util.Locale
@OptIn(ExperimentalMaterial3Api::class)
@Composable
fun MainScreen(
    viewModel: MainViewModel = viewModel(),
    onNewTransactionClicked: () -> Unit = {},
    onLogoutClicked: () -> Unit = {}
) {
    val moeda = stringResource(id = R.string.coin)
    Column(
        modifier = Modifier
            .fillMaxSize()
            .padding(16.dp)
    ) {

        Row(
            modifier = Modifier.fillMaxWidth(),
            horizontalArrangement = Arrangement.SpaceBetween,
            verticalAlignment = Alignment.CenterVertically
        ) {
            Text(text = "${stringResource(id = R.string.greeting)}, ${viewModel.getUserName()}", fontSize = 16.sp, fontWeight = FontWeight.Bold)
            Button(
                colors = ButtonDefaults.buttonColors(Color(0xFF01AA71)),
                onClick = {
                    viewModel.logout()
                    onLogoutClicked()
                },
            ) {
                Text(text = stringResource(id = R.string.logout))
            }
        }
        Spacer(modifier = Modifier.height(16.dp))

        Text(
            modifier = Modifier.align(Alignment.CenterHorizontally),
            text = stringResource(id = R.string.balance), fontSize = 18.sp, fontWeight = FontWeight.Bold, textAlign = TextAlign.Center)

        Spacer(modifier = Modifier.height(8.dp))

        Card(
            modifier = Modifier
                .width(200.dp)
                .align(Alignment.CenterHorizontally),
            shape = RoundedCornerShape(8.dp),
            colors = CardDefaults.cardColors(
                containerColor = Color(0xFF01AA71)
            ),
        ) {
            Text(
                text = "${moeda} ${NumberFormat.getNumberInstance(Locale("pt", "BR")).format(viewModel.userData.planning.balance)}",
                modifier = Modifier
                    .padding(16.dp)
                    .align(Alignment.CenterHorizontally),
                color = Color.White,
                fontSize = 24.sp,
            )
        }

        Spacer(modifier = Modifier.height(16.dp))

        Text(text = stringResource(id = R.string.overview), fontSize = 18.sp, fontWeight = FontWeight.Bold)

        Spacer(modifier = Modifier.height(8.dp))

        LazyColumn(
            modifier = Modifier.weight(1f)
        ) {
            items(viewModel.userData.planning.transactions) { transaction ->
                RecentItem(
                    category = transaction.category + " - " + transaction.description,
                    amount = if (transaction.value < 0) "- ${moeda} ${NumberFormat.getNumberInstance(Locale("pt", "BR")).format(transaction.value + 2*(-transaction.value))}"
                    else "+ ${moeda} ${NumberFormat.getNumberInstance(Locale("pt", "BR")).format(transaction.value)}",
                    date = transaction.date,
                    iconColor = if (transaction.expense) Color.Red else Color.Green
                )
                Spacer(modifier = Modifier.height(8.dp))
            }
        }

        Spacer(modifier = Modifier.height(8.dp))

        Button(
            onClick = { onNewTransactionClicked() },
            modifier = Modifier
                .fillMaxWidth()
                .padding(horizontal = 16.dp),
            colors = ButtonDefaults.buttonColors(Color(0xFF01AA71)),
            shape = RoundedCornerShape(8.dp)
        ) {
            Text(stringResource(id = R.string.new_transaction), color = Color.White)
        }
    }
}

@Composable
fun RecentItem(category: String, amount: String, date: String, iconColor: Color) {
    Row(
        modifier = Modifier
            .fillMaxWidth()
            .padding(8.dp)
    ) {
        Icon(
            imageVector = Icons.Default.ShoppingCart,
            contentDescription = category,
            tint = iconColor,
            modifier = Modifier.size(32.dp)
        )
        Spacer(modifier = Modifier.width(8.dp))
        Column {
            Text(text = category, fontWeight = FontWeight.Bold)
            Text(text = date, color = Color.Gray)
        }
        Spacer(modifier = Modifier.weight(1f))
        Text(text = amount,
            color = if (amount.startsWith("-")) Color.Red else Color.Green)
    }
}

@Preview(showBackground = true)
@Composable
fun MainScreenPreview() {
    WalletAppTheme {
        MainScreen()
    }
}

File: wallet_app/lib/ui/screens/MainScreen.dart
import 'package:flutter/material.dart';
import 'package:intl/intl.dart';
import 'package:provider/provider.dart';
import 'package:wallet_app/repository/dto/planning.dart';
import 'package:wallet_app/repository/dto/user.dart';
import 'package:wallet_app/shared_preferences/user_local_storage.dart';
import 'package:wallet_app/ui/screens/main_view_model.dart';
import 'package:flutter_gen/gen_l10n/app_localizations.dart';

class MainScreen extends StatefulWidget {
  final VoidCallback onNewTransactionClicked;
  final VoidCallback onLogoutClicked;

  MainScreen({
    required this.onNewTransactionClicked,
    required this.onLogoutClicked,
  });

  @override
  _MainScreenState createState() => _MainScreenState();
}

class Transaction {
  final String category;
  final String description;
  final double value;
  final String date;
  final bool expense;

  Transaction({
    required this.category,
    required this.description,
    required this.value,
    required this.date,
    required this.expense,
  });
}


class _MainScreenState extends State<MainScreen> {
  UserLocalStorage repository = UserLocalStorage();
  User _user = User(
    planning: Planning(),
  );
  bool isLoading = true;

  @override
  void initState() {
    super.initState();
    initializeState();

  }

  Future<void> initializeState() async {
    User userStored = await repository.getUser();
    if (userStored.name != "") {
      setState(() {
        _user = userStored;
        isLoading = false;
      });
    } else {
      setState(() {
        isLoading = false;
      });
    }

  }

  @override
  Widget build(BuildContext context) {

    if (isLoading) {
      return Center(child: CircularProgressIndicator());
    }
    final viewModel = Provider.of<MainViewModel>(context, listen: true);
    final user = viewModel.userData.name.length > 0 ? viewModel.userData : _user;
    final transactions = user.transactions;
    final balance = user.balance;
    final moeda = "R\$";

    return Scaffold(
      body: Padding(
        padding: const EdgeInsets.all(16.0),
        child: Column(
          crossAxisAlignment: CrossAxisAlignment.start,
          children: [
            Row(
              mainAxisAlignment: MainAxisAlignment.spaceBetween,
              children: [
                Text(
                  "$AppLocalizations.of(context)!.hello, ${user.fullName}",
                  style: TextStyle(fontSize: 16, fontWeight: FontWeight.bold),
                ),
                ElevatedButton(
                  style: ElevatedButton.styleFrom(
                    backgroundColor: Color(0xFF01AA71),
                  ),
                  onPressed: () {
                    viewModel.logout();
                    Navigator.pushNamed(context, 'login');
                  },
                  child: Text(AppLocalizations.of(context)!.overview),
                ),
              ],
            ),
            SizedBox(height: 16),
            Center(
              child: Text(
                  AppLocalizations.of(context)!.balance,
                style: TextStyle(fontSize: 18, fontWeight: FontWeight.bold),
              ),
            ),
            SizedBox(height: 8),
            Center(
              child: Card(
                color: Color(0xFF01AA71),
                shape: RoundedRectangleBorder(
                  borderRadius: BorderRadius.circular(8),
                ),
                child: Padding(
                  padding: const EdgeInsets.all(16.0),
                  child: Text(
                    "$moeda ${NumberFormat.currency(locale: "pt_BR", symbol: "").format(balance)}",
                    style: TextStyle(color: Colors.white, fontSize: 24),
                  ),
                ),
              ),
            ),
            SizedBox(height: 16),
            Text(
              AppLocalizations.of(context)!.overview,
              style: TextStyle(fontSize: 18, fontWeight: FontWeight.bold),
            ),
            SizedBox(height: 8),
            Expanded(
              child: ListView.builder(
                itemCount: transactions.length,
                itemBuilder: (context, index) {
                  final transaction = transactions[index];
                  return RecentItem(
                    category: "${transaction.category} - ${transaction.description}",
                    amount: transaction.value < 0
                        ? "- $moeda ${NumberFormat.currency(locale: "pt_BR", symbol: "").format(transaction.value + 2 * (-transaction.value))}"
                        : "+ $moeda ${NumberFormat.currency(locale: "pt_BR", symbol: "").format(transaction.value)}",
                    date: transaction.date,
                    iconColor: transaction.expense ? Colors.red : Colors.green,
                  );
                },
              ),
            ),
            SizedBox(height: 8),
            ElevatedButton(
              style: ElevatedButton.styleFrom(
                backgroundColor: Color(0xFF01AA71),
                shape: RoundedRectangleBorder(
                  borderRadius: BorderRadius.circular(8),
                ),
                padding: EdgeInsets.symmetric(horizontal: 16),
              ),
              onPressed: () {Navigator.pushNamed(context, 'transaction'); },
              child: Text(AppLocalizations.of(context)!.newTransaction, style: TextStyle(color: Colors.white)),
            ),
          ],
        ),
      ),
    );
  }
}

class RecentItem extends StatelessWidget {
  final String category;
  final String amount;
  final String date;
  final Color iconColor;

  RecentItem({
    required this.category,
    required this.amount,
    required this.date,
    required this.iconColor,
  });

  @override
  Widget build(BuildContext context) {
    return Padding(
      padding: const EdgeInsets.all(8.0),
      child: Row(
        children: [
          Icon(
            Icons.shopping_cart,
            color: iconColor,
            size: 32,
          ),
          SizedBox(width: 8),
          Expanded(
            child: Column(
              crossAxisAlignment: CrossAxisAlignment.start,
              children: [
                Text(category, style: TextStyle(fontWeight: FontWeight.bold)),
                Text(date, style: TextStyle(color: Colors.grey)),
              ],
            ),
          ),
          Text(
            amount,
            style: TextStyle(color: amount.startsWith("-") ? Colors.red : Colors.green),
          ),
        ],
      ),
    );
  }
}
